diff --git a/BUILD.bazel b/BUILD.bazel
index ac62251e..f4df5be9 100644
--- a/BUILD.bazel
+++ b/BUILD.bazel
@@ -128,7 +128,11 @@ cc_library(
             "-lm",
             "-pthread",
         ],
-        "//conditions:default": ["-pthread"],
+        "//conditions:default": [
+            # NOTE(chokobole): Needed when compiled with --config cuda
+            "-lm",
+            "-pthread"
+        ],
     }),
     deps = select({
         ":has_absl": [
