diff --git a/icicle/benchmarks/curve_benchmarks.cu.cc b/icicle/benchmarks/curve_benchmarks.cu.cc
index 379793ba..707b2c88 100644
--- a/icicle/benchmarks/curve_benchmarks.cu.cc
+++ b/icicle/benchmarks/curve_benchmarks.cu.cc
@@ -1,6 +1,6 @@
 #include <benchmark/benchmark.h>
-#include "utils/test_functions.cuh"
-#include "curves/curve_config.cuh"
+#include "utils/test_functions.cu.h"
+#include "curves/curve_config.cu.h"
 
 using namespace curve_config;
 using namespace benchmark;
diff --git a/icicle/benchmarks/field_benchmarks.cu.cc b/icicle/benchmarks/field_benchmarks.cu.cc
index cf5f2a24..b2d56744 100644
--- a/icicle/benchmarks/field_benchmarks.cu.cc
+++ b/icicle/benchmarks/field_benchmarks.cu.cc
@@ -1,6 +1,6 @@
 #include <benchmark/benchmark.h>
-#include "utils/test_functions.cuh"
-#include "fields/field_config.cuh"
+#include "utils/test_functions.cu.h"
+#include "fields/field_config.cu.h"
 
 using namespace field_config;
 using namespace benchmark;
diff --git a/icicle/include/api/babybear.h b/icicle/include/api/babybear.h
index 64815ceb..e56ccd6e 100644
--- a/icicle/include/api/babybear.h
+++ b/icicle/include/api/babybear.h
@@ -8,13 +8,13 @@
 #define BABYBEAR_API_H
 
 #include <cuda_runtime.h>
-#include "gpu-utils/device_context.cuh"
-#include "merkle-tree/merkle.cuh"
-#include "matrix/matrix.cuh"
-#include "fields/stark_fields/babybear.cuh"
-#include "ntt/ntt.cuh"
-#include "vec_ops/vec_ops.cuh"
-#include "poseidon2/poseidon2.cuh"
+#include "gpu-utils/device_context.cu.h"
+#include "merkle-tree/merkle.cu.h"
+#include "matrix/matrix.cu.h"
+#include "fields/stark_fields/babybear.cu.h"
+#include "ntt/ntt.cu.h"
+#include "vec_ops/vec_ops.cu.h"
+#include "poseidon2/poseidon2.cu.h"
 
 extern "C" cudaError_t babybear_extension_ntt_cuda(
   const babybear::extension_t* input, int size, ntt::NTTDir dir, ntt::NTTConfig<babybear::scalar_t>& config, babybear::extension_t* output);
diff --git a/icicle/include/api/bls12_377.h b/icicle/include/api/bls12_377.h
index 75a113d2..b8cc01bc 100644
--- a/icicle/include/api/bls12_377.h
+++ b/icicle/include/api/bls12_377.h
@@ -8,14 +8,14 @@
 #define BLS12_377_API_H
 
 #include <cuda_runtime.h>
-#include "gpu-utils/device_context.cuh"
-#include "merkle-tree/merkle.cuh"
-#include "matrix/matrix.cuh"
-#include "curves/params/bls12_377.cuh"
-#include "ntt/ntt.cuh"
-#include "msm/msm.cuh"
-#include "vec_ops/vec_ops.cuh"
-#include "poseidon/poseidon.cuh"
+#include "gpu-utils/device_context.cu.h"
+#include "merkle-tree/merkle.cu.h"
+#include "matrix/matrix.cu.h"
+#include "curves/params/bls12_377.cu.h"
+#include "ntt/ntt.cu.h"
+#include "msm/msm.cu.h"
+#include "vec_ops/vec_ops.cu.h"
+#include "poseidon/poseidon.cu.h"
 
 extern "C" cudaError_t bls12_377_g2_precompute_msm_bases_cuda(
   bls12_377::g2_affine_t* bases,
diff --git a/icicle/include/api/bls12_381.h b/icicle/include/api/bls12_381.h
index 9f1a49cd..3814e01a 100644
--- a/icicle/include/api/bls12_381.h
+++ b/icicle/include/api/bls12_381.h
@@ -8,14 +8,14 @@
 #define BLS12_381_API_H
 
 #include <cuda_runtime.h>
-#include "gpu-utils/device_context.cuh"
-#include "merkle-tree/merkle.cuh"
-#include "matrix/matrix.cuh"
-#include "curves/params/bls12_381.cuh"
-#include "ntt/ntt.cuh"
-#include "msm/msm.cuh"
-#include "vec_ops/vec_ops.cuh"
-#include "poseidon/poseidon.cuh"
+#include "gpu-utils/device_context.cu.h"
+#include "merkle-tree/merkle.cu.h"
+#include "matrix/matrix.cu.h"
+#include "curves/params/bls12_381.cu.h"
+#include "ntt/ntt.cu.h"
+#include "msm/msm.cu.h"
+#include "vec_ops/vec_ops.cu.h"
+#include "poseidon/poseidon.cu.h"
 
 extern "C" cudaError_t bls12_381_g2_precompute_msm_bases_cuda(
   bls12_381::g2_affine_t* bases,
diff --git a/icicle/include/api/bn254.h b/icicle/include/api/bn254.h
index 1a9a46b5..16ce4cb4 100644
--- a/icicle/include/api/bn254.h
+++ b/icicle/include/api/bn254.h
@@ -8,15 +8,15 @@
 #define BN254_API_H
 
 #include <cuda_runtime.h>
-#include "gpu-utils/device_context.cuh"
-#include "merkle-tree/merkle.cuh"
-#include "matrix/matrix.cuh"
-#include "curves/params/bn254.cuh"
-#include "ntt/ntt.cuh"
-#include "msm/msm.cuh"
-#include "vec_ops/vec_ops.cuh"
-#include "poseidon/poseidon.cuh"
-#include "poseidon2/poseidon2.cuh"
+#include "gpu-utils/device_context.cu.h"
+#include "merkle-tree/merkle.cu.h"
+#include "matrix/matrix.cu.h"
+#include "curves/params/bn254.cu.h"
+#include "ntt/ntt.cu.h"
+#include "msm/msm.cu.h"
+#include "vec_ops/vec_ops.cu.h"
+#include "poseidon/poseidon.cu.h"
+#include "poseidon2/poseidon2.cu.h"
 
 extern "C" cudaError_t bn254_g2_precompute_msm_bases_cuda(
   bn254::g2_affine_t* bases,
diff --git a/icicle/include/api/bw6_761.h b/icicle/include/api/bw6_761.h
index 531e4514..e311c1cd 100644
--- a/icicle/include/api/bw6_761.h
+++ b/icicle/include/api/bw6_761.h
@@ -8,14 +8,14 @@
 #define BW6_761_API_H
 
 #include <cuda_runtime.h>
-#include "gpu-utils/device_context.cuh"
-#include "merkle-tree/merkle.cuh"
-#include "matrix/matrix.cuh"
-#include "curves/params/bw6_761.cuh"
-#include "ntt/ntt.cuh"
-#include "msm/msm.cuh"
-#include "vec_ops/vec_ops.cuh"
-#include "poseidon/poseidon.cuh"
+#include "gpu-utils/device_context.cu.h"
+#include "merkle-tree/merkle.cu.h"
+#include "matrix/matrix.cu.h"
+#include "curves/params/bw6_761.cu.h"
+#include "ntt/ntt.cu.h"
+#include "msm/msm.cu.h"
+#include "vec_ops/vec_ops.cu.h"
+#include "poseidon/poseidon.cu.h"
 
 extern "C" cudaError_t bw6_761_g2_precompute_msm_bases_cuda(
   bw6_761::g2_affine_t* bases,
diff --git a/icicle/include/api/grumpkin.h b/icicle/include/api/grumpkin.h
index 40241e59..c9b302f4 100644
--- a/icicle/include/api/grumpkin.h
+++ b/icicle/include/api/grumpkin.h
@@ -8,13 +8,13 @@
 #define GRUMPKIN_API_H
 
 #include <cuda_runtime.h>
-#include "gpu-utils/device_context.cuh"
-#include "merkle-tree/merkle.cuh"
-#include "matrix/matrix.cuh"
-#include "curves/params/grumpkin.cuh"
-#include "msm/msm.cuh"
-#include "vec_ops/vec_ops.cuh"
-#include "poseidon/poseidon.cuh"
+#include "gpu-utils/device_context.cu.h"
+#include "merkle-tree/merkle.cu.h"
+#include "matrix/matrix.cu.h"
+#include "curves/params/grumpkin.cu.h"
+#include "msm/msm.cu.h"
+#include "vec_ops/vec_ops.cu.h"
+#include "poseidon/poseidon.cu.h"
 
 extern "C" cudaError_t grumpkin_precompute_msm_bases_cuda(
   grumpkin::affine_t* bases,
diff --git a/icicle/include/api/hash.h b/icicle/include/api/hash.h
index ffa40d81..2c9436e2 100644
--- a/icicle/include/api/hash.h
+++ b/icicle/include/api/hash.h
@@ -4,9 +4,9 @@
 #define HASH_API_H
 
 #include <cuda_runtime.h>
-#include "gpu-utils/device_context.cuh"
-#include "hash/keccak/keccak.cuh"
-#include "merkle-tree/merkle.cuh"
+#include "gpu-utils/device_context.cu.h"
+#include "hash/keccak/keccak.cu.h"
+#include "merkle-tree/merkle.cu.h"
 
 extern "C" cudaError_t
   keccak256_cuda(uint8_t* input, int input_block_size, int number_of_blocks, uint8_t* output, keccak::HashConfig& config);
diff --git a/icicle/include/api/m31.h b/icicle/include/api/m31.h
index 277268a1..19879db0 100644
--- a/icicle/include/api/m31.h
+++ b/icicle/include/api/m31.h
@@ -8,11 +8,11 @@
 #define M31_API_H
 
 #include <cuda_runtime.h>
-#include "gpu-utils/device_context.cuh"
-#include "merkle-tree/merkle.cuh"
-#include "matrix/matrix.cuh"
-#include "fields/stark_fields/m31.cuh"
-#include "vec_ops/vec_ops.cuh"
+#include "gpu-utils/device_context.cu.h"
+#include "merkle-tree/merkle.cu.h"
+#include "matrix/matrix.cu.h"
+#include "fields/stark_fields/m31.cu.h"
+#include "vec_ops/vec_ops.cu.h"
 
 extern "C" cudaError_t m31_build_merkle_tree(
   const m31::scalar_t* leaves,
diff --git a/icicle/include/api/stark252.h b/icicle/include/api/stark252.h
index 867f98e6..24f8dbd5 100644
--- a/icicle/include/api/stark252.h
+++ b/icicle/include/api/stark252.h
@@ -8,12 +8,12 @@
 #define STARK252_API_H
 
 #include <cuda_runtime.h>
-#include "gpu-utils/device_context.cuh"
-#include "merkle-tree/merkle.cuh"
-#include "matrix/matrix.cuh"
-#include "fields/stark_fields/stark252.cuh"
-#include "ntt/ntt.cuh"
-#include "vec_ops/vec_ops.cuh"
+#include "gpu-utils/device_context.cu.h"
+#include "merkle-tree/merkle.cu.h"
+#include "matrix/matrix.cu.h"
+#include "fields/stark_fields/stark252.cu.h"
+#include "ntt/ntt.cu.h"
+#include "vec_ops/vec_ops.cu.h"
 
 extern "C" cudaError_t stark252_build_merkle_tree(
   const stark252::scalar_t* leaves,
diff --git a/icicle/include/curves/affine.cu.h b/icicle/include/curves/affine.cu.h
index 2b1253e4..c9ddb5c2 100644
--- a/icicle/include/curves/affine.cu.h
+++ b/icicle/include/curves/affine.cu.h
@@ -1,7 +1,7 @@
 #pragma once
 
-#include "gpu-utils/sharedmem.cuh"
-#include "gpu-utils/modifiers.cuh"
+#include "gpu-utils/sharedmem.cu.h"
+#include "gpu-utils/modifiers.cu.h"
 #include <iostream>
 
 template <class FF>
diff --git a/icicle/include/curves/curve_config.cu.h b/icicle/include/curves/curve_config.cu.h
index c9fe109b..fb219ca2 100644
--- a/icicle/include/curves/curve_config.cu.h
+++ b/icicle/include/curves/curve_config.cu.h
@@ -3,7 +3,7 @@
 #define CURVE_CONFIG_H
 
 #include "fields/id.h"
-#include "curves/projective.cuh"
+#include "curves/projective.cu.h"
 
 /**
  * @namespace curve_config
@@ -12,23 +12,23 @@
  * with the `-DCURVE` env variable passed during build.
  */
 #if CURVE_ID == BN254
-#include "curves/params/bn254.cuh"
+#include "curves/params/bn254.cu.h"
 namespace curve_config = bn254;
 
 #elif CURVE_ID == BLS12_381
-#include "curves/params/bls12_381.cuh"
+#include "curves/params/bls12_381.cu.h"
 namespace curve_config = bls12_381;
 
 #elif CURVE_ID == BLS12_377
-#include "curves/params/bls12_377.cuh"
+#include "curves/params/bls12_377.cu.h"
 namespace curve_config = bls12_377;
 
 #elif CURVE_ID == BW6_761
-#include "curves/params/bw6_761.cuh"
+#include "curves/params/bw6_761.cu.h"
 namespace curve_config = bw6_761;
 
 #elif CURVE_ID == GRUMPKIN
-#include "curves/params/grumpkin.cuh"
+#include "curves/params/grumpkin.cu.h"
 namespace curve_config = grumpkin;
 #endif
 #endif
\ No newline at end of file
diff --git a/icicle/include/curves/params/bls12_377.cu.h b/icicle/include/curves/params/bls12_377.cu.h
index 3e81fb17..5c01bcec 100644
--- a/icicle/include/curves/params/bls12_377.cu.h
+++ b/icicle/include/curves/params/bls12_377.cu.h
@@ -2,13 +2,13 @@
 #ifndef BLS12_377_PARAMS_H
 #define BLS12_377_PARAMS_H
 
-#include "fields/storage.cuh"
+#include "fields/storage.cu.h"
 
 #include "curves/macro.h"
-#include "curves/projective.cuh"
-#include "fields/snark_fields/bls12_377_base.cuh"
-#include "fields/snark_fields/bls12_377_scalar.cuh"
-#include "fields/quadratic_extension.cuh"
+#include "curves/projective.cu.h"
+#include "fields/snark_fields/bls12_377_base.cu.h"
+#include "fields/snark_fields/bls12_377_scalar.cu.h"
+#include "fields/quadratic_extension.cu.h"
 
 namespace bls12_377 {
   // G1 and G2 generators
diff --git a/icicle/include/curves/params/bls12_381.cu.h b/icicle/include/curves/params/bls12_381.cu.h
index 86c7ef6a..d8998b70 100644
--- a/icicle/include/curves/params/bls12_381.cu.h
+++ b/icicle/include/curves/params/bls12_381.cu.h
@@ -2,13 +2,13 @@
 #ifndef BLS12_381_PARAMS_H
 #define BLS12_381_PARAMS_H
 
-#include "fields/storage.cuh"
+#include "fields/storage.cu.h"
 
 #include "curves/macro.h"
-#include "curves/projective.cuh"
-#include "fields/snark_fields/bls12_381_base.cuh"
-#include "fields/snark_fields/bls12_381_scalar.cuh"
-#include "fields/quadratic_extension.cuh"
+#include "curves/projective.cu.h"
+#include "fields/snark_fields/bls12_381_base.cu.h"
+#include "fields/snark_fields/bls12_381_scalar.cu.h"
+#include "fields/quadratic_extension.cu.h"
 
 namespace bls12_381 {
   // G1 and G2 generators
diff --git a/icicle/include/curves/params/bn254.cu.h b/icicle/include/curves/params/bn254.cu.h
index 4ae43760..88dedd3d 100644
--- a/icicle/include/curves/params/bn254.cu.h
+++ b/icicle/include/curves/params/bn254.cu.h
@@ -2,13 +2,13 @@
 #ifndef BN254_PARAMS_H
 #define BN254_PARAMS_H
 
-#include "fields/storage.cuh"
+#include "fields/storage.cu.h"
 
 #include "curves/macro.h"
-#include "curves/projective.cuh"
-#include "fields/snark_fields/bn254_base.cuh"
-#include "fields/snark_fields/bn254_scalar.cuh"
-#include "fields/quadratic_extension.cuh"
+#include "curves/projective.cu.h"
+#include "fields/snark_fields/bn254_base.cu.h"
+#include "fields/snark_fields/bn254_scalar.cu.h"
+#include "fields/quadratic_extension.cu.h"
 
 namespace bn254 {
   // G1 and G2 generators
diff --git a/icicle/include/curves/params/bw6_761.cu.h b/icicle/include/curves/params/bw6_761.cu.h
index ce649f5e..48898dce 100644
--- a/icicle/include/curves/params/bw6_761.cu.h
+++ b/icicle/include/curves/params/bw6_761.cu.h
@@ -2,13 +2,13 @@
 #ifndef BW6_761_PARAMS_H
 #define BW6_761_PARAMS_H
 
-#include "fields/storage.cuh"
+#include "fields/storage.cu.h"
 
 #include "curves/macro.h"
-#include "curves/projective.cuh"
-#include "fields/snark_fields/bw6_761_base.cuh"
-#include "fields/snark_fields/bw6_761_scalar.cuh"
-#include "fields/quadratic_extension.cuh"
+#include "curves/projective.cu.h"
+#include "fields/snark_fields/bw6_761_base.cu.h"
+#include "fields/snark_fields/bw6_761_scalar.cu.h"
+#include "fields/quadratic_extension.cu.h"
 
 namespace bw6_761 {
   // G1 and G2 generators
diff --git a/icicle/include/curves/params/grumpkin.cu.h b/icicle/include/curves/params/grumpkin.cu.h
index 855897a1..13c9e11f 100644
--- a/icicle/include/curves/params/grumpkin.cu.h
+++ b/icicle/include/curves/params/grumpkin.cu.h
@@ -2,12 +2,12 @@
 #ifndef GRUMPKIN_PARAMS_H
 #define GRUMPKIN_PARAMS_H
 
-#include "fields/storage.cuh"
+#include "fields/storage.cu.h"
 
 #include "curves/macro.h"
-#include "curves/projective.cuh"
-#include "fields/snark_fields/grumpkin_base.cuh"
-#include "fields/snark_fields/grumpkin_scalar.cuh"
+#include "curves/projective.cu.h"
+#include "fields/snark_fields/grumpkin_base.cu.h"
+#include "fields/snark_fields/grumpkin_scalar.cu.h"
 
 namespace grumpkin {
   typedef bn254::fp_config fq_config;
diff --git a/icicle/include/curves/projective.cu.h b/icicle/include/curves/projective.cu.h
index 47ff27a8..6d08460c 100644
--- a/icicle/include/curves/projective.cu.h
+++ b/icicle/include/curves/projective.cu.h
@@ -1,7 +1,7 @@
 #pragma once
 
-#include "affine.cuh"
-#include "gpu-utils/sharedmem.cuh"
+#include "affine.cu.h"
+#include "gpu-utils/sharedmem.cu.h"
 
 template <typename FF, class SCALAR_FF, const FF& B_VALUE, const FF& GENERATOR_X, const FF& GENERATOR_Y>
 class Projective
diff --git a/icicle/include/fields/field.cu.h b/icicle/include/fields/field.cu.h
index 6fe3e069..fbde1b33 100644
--- a/icicle/include/fields/field.cu.h
+++ b/icicle/include/fields/field.cu.h
@@ -18,12 +18,12 @@
 
 #pragma once
 
-#include "gpu-utils/error_handler.cuh"
-#include "gpu-utils/modifiers.cuh"
-#include "gpu-utils/sharedmem.cuh"
-#include "host_math.cuh"
-#include "ptx.cuh"
-#include "storage.cuh"
+#include "gpu-utils/error_handler.cu.h"
+#include "gpu-utils/modifiers.cu.h"
+#include "gpu-utils/sharedmem.cu.h"
+#include "host_math.cu.h"
+#include "ptx.cu.h"
+#include "storage.cu.h"
 
 #include <iomanip>
 #include <iostream>
diff --git a/icicle/include/fields/field_config.cu.h b/icicle/include/fields/field_config.cu.h
index d9ec18c0..72e1250e 100644
--- a/icicle/include/fields/field_config.cu.h
+++ b/icicle/include/fields/field_config.cu.h
@@ -3,7 +3,7 @@
 #define FIELD_CONFIG_H
 
 #include "fields/id.h"
-#include "fields/field.cuh"
+#include "fields/field.cu.h"
 
 /**
  * @namespace field_config
@@ -11,30 +11,30 @@
  * with the `-DFIELD` env variable passed during build.
  */
 #if FIELD_ID == BN254
-#include "fields/snark_fields/bn254_scalar.cuh"
+#include "fields/snark_fields/bn254_scalar.cu.h"
 namespace field_config = bn254;
 #elif FIELD_ID == BLS12_381
-#include "fields/snark_fields/bls12_381_scalar.cuh"
+#include "fields/snark_fields/bls12_381_scalar.cu.h"
 using bls12_381::fp_config;
 namespace field_config = bls12_381;
 #elif FIELD_ID == BLS12_377
-#include "fields/snark_fields/bls12_377_scalar.cuh"
+#include "fields/snark_fields/bls12_377_scalar.cu.h"
 namespace field_config = bls12_377;
 #elif FIELD_ID == BW6_761
-#include "fields/snark_fields/bw6_761_scalar.cuh"
+#include "fields/snark_fields/bw6_761_scalar.cu.h"
 namespace field_config = bw6_761;
 #elif FIELD_ID == GRUMPKIN
-#include "fields/snark_fields/grumpkin_scalar.cuh"
+#include "fields/snark_fields/grumpkin_scalar.cu.h"
 namespace field_config = grumpkin;
 
 #elif FIELD_ID == BABY_BEAR
-#include "fields/stark_fields/babybear.cuh"
+#include "fields/stark_fields/babybear.cu.h"
 namespace field_config = babybear;
 #elif FIELD_ID == STARK_252
-#include "fields/stark_fields/stark252.cuh"
+#include "fields/stark_fields/stark252.cu.h"
 namespace field_config = stark252;
 #elif FIELD_ID == M31
-#include "fields/stark_fields/m31.cuh"
+#include "fields/stark_fields/m31.cu.h"
 namespace field_config = m31;
 #endif
 
diff --git a/icicle/include/fields/host_math.cu.h b/icicle/include/fields/host_math.cu.h
index 0cdde0ca..7669fd42 100644
--- a/icicle/include/fields/host_math.cu.h
+++ b/icicle/include/fields/host_math.cu.h
@@ -4,8 +4,8 @@
 
 #include <cstdint>
 #include <cuda_runtime.h>
-#include "gpu-utils/modifiers.cuh"
-#include "storage.cuh"
+#include "gpu-utils/modifiers.cu.h"
+#include "storage.cu.h"
 
 namespace host_math {
 
diff --git a/icicle/include/fields/params_gen.cu.h b/icicle/include/fields/params_gen.cu.h
index 6b21c5e0..638a9391 100644
--- a/icicle/include/fields/params_gen.cu.h
+++ b/icicle/include/fields/params_gen.cu.h
@@ -2,8 +2,8 @@
 #ifndef PARAMS_GEN_H
 #define PARAMS_GEN_H
 
-#include "storage.cuh"
-#include "host_math.cuh"
+#include "storage.cu.h"
+#include "host_math.cu.h"
 
 namespace params_gen {
   template <unsigned NLIMBS, unsigned BIT_SHIFT>
diff --git a/icicle/include/fields/quadratic_extension.cu.h b/icicle/include/fields/quadratic_extension.cu.h
index 10065386..6c4f0dff 100644
--- a/icicle/include/fields/quadratic_extension.cu.h
+++ b/icicle/include/fields/quadratic_extension.cu.h
@@ -1,8 +1,8 @@
 #pragma once
 
-#include "field.cuh"
-#include "gpu-utils/modifiers.cuh"
-#include "gpu-utils/sharedmem.cuh"
+#include "field.cu.h"
+#include "gpu-utils/modifiers.cu.h"
+#include "gpu-utils/sharedmem.cu.h"
 
 template <typename CONFIG, class T>
 class ExtensionField
diff --git a/icicle/include/fields/quartic_extension.cu.h b/icicle/include/fields/quartic_extension.cu.h
index 8fead58c..609ecf36 100644
--- a/icicle/include/fields/quartic_extension.cu.h
+++ b/icicle/include/fields/quartic_extension.cu.h
@@ -1,8 +1,8 @@
 #pragma once
 
-#include "field.cuh"
-#include "gpu-utils/modifiers.cuh"
-#include "gpu-utils/sharedmem.cuh"
+#include "field.cu.h"
+#include "gpu-utils/modifiers.cu.h"
+#include "gpu-utils/sharedmem.cu.h"
 
 template <typename CONFIG, class T>
 class ExtensionField
diff --git a/icicle/include/fields/snark_fields/bls12_377_base.cu.h b/icicle/include/fields/snark_fields/bls12_377_base.cu.h
index dc7a13fe..abf9b89f 100644
--- a/icicle/include/fields/snark_fields/bls12_377_base.cu.h
+++ b/icicle/include/fields/snark_fields/bls12_377_base.cu.h
@@ -2,8 +2,8 @@
 #ifndef BLS12_377_BASE_PARAMS_H
 #define BLS12_377_BASE_PARAMS_H
 
-#include "fields/storage.cuh"
-#include "fields/params_gen.cuh"
+#include "fields/storage.cu.h"
+#include "fields/params_gen.cu.h"
 
 namespace bls12_377 {
   struct fq_config {
diff --git a/icicle/include/fields/snark_fields/bls12_377_scalar.cu.h b/icicle/include/fields/snark_fields/bls12_377_scalar.cu.h
index 3924f59f..0b0ded12 100644
--- a/icicle/include/fields/snark_fields/bls12_377_scalar.cu.h
+++ b/icicle/include/fields/snark_fields/bls12_377_scalar.cu.h
@@ -2,9 +2,9 @@
 #ifndef BLS12_377_SCALAR_PARAMS_H
 #define BLS12_377_SCALAR_PARAMS_H
 
-#include "fields/storage.cuh"
-#include "fields/field.cuh"
-#include "fields/params_gen.cuh"
+#include "fields/storage.cu.h"
+#include "fields/field.cu.h"
+#include "fields/params_gen.cu.h"
 
 namespace bls12_377 {
   struct fp_config {
diff --git a/icicle/include/fields/snark_fields/bls12_381_base.cu.h b/icicle/include/fields/snark_fields/bls12_381_base.cu.h
index 54fc0ab5..cef89c40 100644
--- a/icicle/include/fields/snark_fields/bls12_381_base.cu.h
+++ b/icicle/include/fields/snark_fields/bls12_381_base.cu.h
@@ -2,8 +2,8 @@
 #ifndef BLS12_381_BASE_PARAMS_H
 #define BLS12_381_BASE_PARAMS_H
 
-#include "fields/storage.cuh"
-#include "fields/params_gen.cuh"
+#include "fields/storage.cu.h"
+#include "fields/params_gen.cu.h"
 
 namespace bls12_381 {
   struct fq_config {
diff --git a/icicle/include/fields/snark_fields/bls12_381_scalar.cu.h b/icicle/include/fields/snark_fields/bls12_381_scalar.cu.h
index c6bd12fe..61e9f8bc 100644
--- a/icicle/include/fields/snark_fields/bls12_381_scalar.cu.h
+++ b/icicle/include/fields/snark_fields/bls12_381_scalar.cu.h
@@ -2,9 +2,9 @@
 #ifndef BLS12_381_SCALAR_PARAMS_H
 #define BLS12_381_SCALAR_PARAMS_H
 
-#include "fields/storage.cuh"
-#include "fields/field.cuh"
-#include "fields/params_gen.cuh"
+#include "fields/storage.cu.h"
+#include "fields/field.cu.h"
+#include "fields/params_gen.cu.h"
 
 namespace bls12_381 {
   struct fp_config {
diff --git a/icicle/include/fields/snark_fields/bn254_base.cu.h b/icicle/include/fields/snark_fields/bn254_base.cu.h
index 95f18ed3..da993cbd 100644
--- a/icicle/include/fields/snark_fields/bn254_base.cu.h
+++ b/icicle/include/fields/snark_fields/bn254_base.cu.h
@@ -2,8 +2,8 @@
 #ifndef BN254_BASE_PARAMS_H
 #define BN254_BASE_PARAMS_H
 
-#include "fields/storage.cuh"
-#include "fields/params_gen.cuh"
+#include "fields/storage.cu.h"
+#include "fields/params_gen.cu.h"
 
 namespace bn254 {
   struct fq_config {
diff --git a/icicle/include/fields/snark_fields/bn254_scalar.cu.h b/icicle/include/fields/snark_fields/bn254_scalar.cu.h
index 67478e2d..8911357c 100644
--- a/icicle/include/fields/snark_fields/bn254_scalar.cu.h
+++ b/icicle/include/fields/snark_fields/bn254_scalar.cu.h
@@ -2,9 +2,9 @@
 #ifndef BN254_SCALAR_PARAMS_H
 #define BN254_SCALAR_PARAMS_H
 
-#include "fields/storage.cuh"
-#include "fields/field.cuh"
-#include "fields/params_gen.cuh"
+#include "fields/storage.cu.h"
+#include "fields/field.cu.h"
+#include "fields/params_gen.cu.h"
 
 namespace bn254 {
   struct fp_config {
diff --git a/icicle/include/fields/snark_fields/bw6_761_base.cu.h b/icicle/include/fields/snark_fields/bw6_761_base.cu.h
index 4ec110ff..6ab52e19 100644
--- a/icicle/include/fields/snark_fields/bw6_761_base.cu.h
+++ b/icicle/include/fields/snark_fields/bw6_761_base.cu.h
@@ -2,8 +2,8 @@
 #ifndef BW6_761_BASE_BASE_H
 #define BW6_761_BASE_BASE_H
 
-#include "fields/storage.cuh"
-#include "fields/params_gen.cuh"
+#include "fields/storage.cu.h"
+#include "fields/params_gen.cu.h"
 
 namespace bw6_761 {
   struct fq_config {
diff --git a/icicle/include/fields/snark_fields/bw6_761_scalar.cu.h b/icicle/include/fields/snark_fields/bw6_761_scalar.cu.h
index 304595a7..cb9205ab 100644
--- a/icicle/include/fields/snark_fields/bw6_761_scalar.cu.h
+++ b/icicle/include/fields/snark_fields/bw6_761_scalar.cu.h
@@ -2,9 +2,9 @@
 #ifndef BW6_761_SCALAR_PARAMS_H
 #define BW6_761_SCALAR_PARAMS_H
 
-#include "fields/storage.cuh"
-#include "fields/field.cuh"
-#include "fields/snark_fields/bls12_377_base.cuh"
+#include "fields/storage.cu.h"
+#include "fields/field.cu.h"
+#include "fields/snark_fields/bls12_377_base.cu.h"
 
 namespace bw6_761 {
   typedef bls12_377::fq_config fp_config;
diff --git a/icicle/include/fields/snark_fields/grumpkin_base.cu.h b/icicle/include/fields/snark_fields/grumpkin_base.cu.h
index 355f2f30..95d7ed3e 100644
--- a/icicle/include/fields/snark_fields/grumpkin_base.cu.h
+++ b/icicle/include/fields/snark_fields/grumpkin_base.cu.h
@@ -2,8 +2,8 @@
 #ifndef GRUMPKIN_BASE_PARAMS_H
 #define GRUMPKIN_BASE_PARAMS_H
 
-#include "fields/storage.cuh"
-#include "fields/snark_fields/bn254_scalar.cuh"
+#include "fields/storage.cu.h"
+#include "fields/snark_fields/bn254_scalar.cu.h"
 
 namespace grumpkin {
   typedef bn254::fp_config fq_config;
diff --git a/icicle/include/fields/snark_fields/grumpkin_scalar.cu.h b/icicle/include/fields/snark_fields/grumpkin_scalar.cu.h
index 4354c8de..87287e1b 100644
--- a/icicle/include/fields/snark_fields/grumpkin_scalar.cu.h
+++ b/icicle/include/fields/snark_fields/grumpkin_scalar.cu.h
@@ -2,9 +2,9 @@
 #ifndef GRUMPKIN_SCALAR_PARAMS_H
 #define GRUMPKIN_SCALAR_PARAMS_H
 
-#include "fields/storage.cuh"
-#include "fields/field.cuh"
-#include "fields/snark_fields/bn254_base.cuh"
+#include "fields/storage.cu.h"
+#include "fields/field.cu.h"
+#include "fields/snark_fields/bn254_base.cu.h"
 
 namespace grumpkin {
   typedef bn254::fq_config fp_config;
diff --git a/icicle/include/fields/stark_fields/babybear.cu.h b/icicle/include/fields/stark_fields/babybear.cu.h
index 57f3f06e..76ea9691 100644
--- a/icicle/include/fields/stark_fields/babybear.cu.h
+++ b/icicle/include/fields/stark_fields/babybear.cu.h
@@ -1,9 +1,9 @@
 #pragma once
 
-#include "fields/storage.cuh"
-#include "fields/field.cuh"
-#include "fields/quartic_extension.cuh"
-#include "fields/params_gen.cuh"
+#include "fields/storage.cu.h"
+#include "fields/field.cu.h"
+#include "fields/quartic_extension.cu.h"
+#include "fields/params_gen.cu.h"
 
 namespace babybear {
   struct fp_config {
diff --git a/icicle/include/fields/stark_fields/m31.cu.h b/icicle/include/fields/stark_fields/m31.cu.h
index b45592ce..b7fe5e5d 100644
--- a/icicle/include/fields/stark_fields/m31.cu.h
+++ b/icicle/include/fields/stark_fields/m31.cu.h
@@ -1,8 +1,8 @@
 #pragma once
 
-#include "fields/storage.cuh"
-#include "fields/field.cuh"
-#include "fields/quartic_extension.cuh"
+#include "fields/storage.cu.h"
+#include "fields/field.cu.h"
+#include "fields/quartic_extension.cu.h"
 
 namespace m31 {
   template <class CONFIG>
diff --git a/icicle/include/fields/stark_fields/stark252.cu.h b/icicle/include/fields/stark_fields/stark252.cu.h
index d61f757e..6e45386c 100644
--- a/icicle/include/fields/stark_fields/stark252.cu.h
+++ b/icicle/include/fields/stark_fields/stark252.cu.h
@@ -1,8 +1,8 @@
 #pragma once
 
-#include "fields/storage.cuh"
-#include "fields/field.cuh"
-#include "fields/params_gen.cuh"
+#include "fields/storage.cu.h"
+#include "fields/field.cu.h"
+#include "fields/params_gen.cu.h"
 
 // modulus = 3618502788666131213697322783095070105623107215331596699973092056135872020481 (2^251+17*2^192+1)
 namespace stark252 {
diff --git a/icicle/include/gpu-utils/device_context.cu.h b/icicle/include/gpu-utils/device_context.cu.h
index b1448a5a..31658426 100644
--- a/icicle/include/gpu-utils/device_context.cu.h
+++ b/icicle/include/gpu-utils/device_context.cu.h
@@ -3,7 +3,7 @@
 #define DEVICE_CONTEXT_H
 
 #include <cuda_runtime.h>
-#include "gpu-utils/error_handler.cuh"
+#include "gpu-utils/error_handler.cu.h"
 
 namespace device_context {
 
diff --git a/icicle/include/hash/hash.cu.h b/icicle/include/hash/hash.cu.h
index 338d337d..7ca414cc 100644
--- a/icicle/include/hash/hash.cu.h
+++ b/icicle/include/hash/hash.cu.h
@@ -2,9 +2,9 @@
 #ifndef HASH_H
 #define HASH_H
 
-#include "gpu-utils/device_context.cuh"
-#include "gpu-utils/error_handler.cuh"
-#include "matrix/matrix.cuh"
+#include "gpu-utils/device_context.cu.h"
+#include "gpu-utils/error_handler.cu.h"
+#include "matrix/matrix.cu.h"
 #include <cassert>
 
 using matrix::Matrix;
diff --git a/icicle/include/hash/keccak/keccak.cu.h b/icicle/include/hash/keccak/keccak.cu.h
index d95c22a4..eee2f806 100644
--- a/icicle/include/hash/keccak/keccak.cu.h
+++ b/icicle/include/hash/keccak/keccak.cu.h
@@ -3,10 +3,10 @@
 #define KECCAK_H
 
 #include <cstdint>
-#include "gpu-utils/device_context.cuh"
-#include "gpu-utils/error_handler.cuh"
+#include "gpu-utils/device_context.cu.h"
+#include "gpu-utils/error_handler.cu.h"
 
-#include "hash/hash.cuh"
+#include "hash/hash.cu.h"
 
 using namespace hash;
 
diff --git a/icicle/include/merkle-tree/merkle.cu.h b/icicle/include/merkle-tree/merkle.cu.h
index ba50ff76..d0a020db 100644
--- a/icicle/include/merkle-tree/merkle.cu.h
+++ b/icicle/include/merkle-tree/merkle.cu.h
@@ -2,11 +2,11 @@
 #ifndef MERKLE_H
 #define MERKLE_H
 
-#include "gpu-utils/device_context.cuh"
-#include "gpu-utils/error_handler.cuh"
+#include "gpu-utils/device_context.cu.h"
+#include "gpu-utils/error_handler.cu.h"
 #include "utils/utils.h"
-#include "hash/hash.cuh"
-#include "matrix/matrix.cuh"
+#include "hash/hash.cu.h"
+#include "matrix/matrix.cu.h"
 
 #include <vector>
 #include <numeric>
diff --git a/icicle/include/msm/msm.cu.h b/icicle/include/msm/msm.cu.h
index 1bd27a48..e393799b 100644
--- a/icicle/include/msm/msm.cu.h
+++ b/icicle/include/msm/msm.cu.h
@@ -4,11 +4,11 @@
 
 #include <cuda_runtime.h>
 
-#include "curves/affine.cuh"
-#include "curves/projective.cuh"
-#include "fields/field.cuh"
-#include "gpu-utils/device_context.cuh"
-#include "gpu-utils/error_handler.cuh"
+#include "curves/affine.cu.h"
+#include "curves/projective.cu.h"
+#include "fields/field.cu.h"
+#include "gpu-utils/device_context.cu.h"
+#include "gpu-utils/error_handler.cu.h"
 
 /**
  * @namespace msm
diff --git a/icicle/include/ntt/ntt.cu.h b/icicle/include/ntt/ntt.cu.h
index 1f032bb4..65482d85 100644
--- a/icicle/include/ntt/ntt.cu.h
+++ b/icicle/include/ntt/ntt.cu.h
@@ -4,10 +4,10 @@
 
 #include <cuda_runtime.h>
 
-#include "gpu-utils/device_context.cuh"
-#include "gpu-utils/error_handler.cuh"
-#include "gpu-utils/sharedmem.cuh"
-#include "utils/utils_kernels.cuh"
+#include "gpu-utils/device_context.cu.h"
+#include "gpu-utils/error_handler.cu.h"
+#include "gpu-utils/sharedmem.cu.h"
+#include "utils/utils_kernels.cu.h"
 #include "utils/utils.h"
 
 /**
diff --git a/icicle/include/ntt/ntt_impl.cu.h b/icicle/include/ntt/ntt_impl.cu.h
index 5f5bf1e2..56b33bb8 100644
--- a/icicle/include/ntt/ntt_impl.cu.h
+++ b/icicle/include/ntt/ntt_impl.cu.h
@@ -3,7 +3,7 @@
 #define _NTT_IMPL_H
 
 #include <stdint.h>
-#include "ntt/ntt.cuh" // for enum Ordering
+#include "ntt/ntt.cu.h" // for enum Ordering
 
 namespace mxntt {
 
diff --git a/icicle/include/polynomials/cuda_backend/polynomial_cuda_backend.cu.h b/icicle/include/polynomials/cuda_backend/polynomial_cuda_backend.cu.h
index da07254a..40d7bce6 100644
--- a/icicle/include/polynomials/cuda_backend/polynomial_cuda_backend.cu.h
+++ b/icicle/include/polynomials/cuda_backend/polynomial_cuda_backend.cu.h
@@ -1,7 +1,7 @@
 #pragma once
 
-#include "gpu-utils/device_context.cuh"
-#include "fields/field_config.cuh"
+#include "gpu-utils/device_context.cu.h"
+#include "fields/field_config.cu.h"
 #include "polynomials/polynomials.h"
 
 using device_context::DeviceContext;
diff --git a/icicle/include/polynomials/polynomials.h b/icicle/include/polynomials/polynomials.h
index 7916b837..f61f794f 100644
--- a/icicle/include/polynomials/polynomials.h
+++ b/icicle/include/polynomials/polynomials.h
@@ -3,7 +3,7 @@
 #include <iostream>
 #include <memory>
 #include "utils/integrity_pointer.h"
-#include "fields/field_config.cuh"
+#include "fields/field_config.cu.h"
 
 #include "polynomial_context.h"
 #include "polynomial_backend.h"
diff --git a/icicle/include/poseidon/kernels.cu.h b/icicle/include/poseidon/kernels.cu.h
index 4afa9878..b9ca6972 100644
--- a/icicle/include/poseidon/kernels.cu.h
+++ b/icicle/include/poseidon/kernels.cu.h
@@ -2,8 +2,8 @@
 #ifndef POSEIDON_KERNELS_H
 #define POSEIDON_KERNELS_H
 
-#include "gpu-utils/modifiers.cuh"
-#include "poseidon/constants.cuh"
+#include "gpu-utils/modifiers.cu.h"
+#include "poseidon/constants.cu.h"
 
 namespace poseidon {
   template <typename S, int T>
diff --git a/icicle/include/poseidon/poseidon.cu.h b/icicle/include/poseidon/poseidon.cu.h
index 24727c21..3b1d6b3b 100644
--- a/icicle/include/poseidon/poseidon.cu.h
+++ b/icicle/include/poseidon/poseidon.cu.h
@@ -4,13 +4,13 @@
 
 #include <cstdint>
 #include <stdexcept>
-#include "gpu-utils/device_context.cuh"
-#include "gpu-utils/error_handler.cuh"
+#include "gpu-utils/device_context.cu.h"
+#include "gpu-utils/error_handler.cu.h"
 #include "utils/utils.h"
 
-#include "poseidon/kernels.cuh"
-#include "poseidon/constants.cuh"
-#include "hash/hash.cuh"
+#include "poseidon/kernels.cu.h"
+#include "poseidon/constants.cu.h"
+#include "hash/hash.cu.h"
 using namespace hash;
 
 /**
diff --git a/icicle/include/poseidon2/constants.cu.h b/icicle/include/poseidon2/constants.cu.h
index 35bc2ebf..5b20ef6b 100644
--- a/icicle/include/poseidon2/constants.cu.h
+++ b/icicle/include/poseidon2/constants.cu.h
@@ -2,7 +2,7 @@
 #ifndef POSEIDON2_CONSTANTS_H
 #define POSEIDON2_CONSTANTS_H
 
-#include "gpu-utils/device_context.cuh"
+#include "gpu-utils/device_context.cu.h"
 
 namespace poseidon2 {
   /**
diff --git a/icicle/include/poseidon2/kernels.cu.h b/icicle/include/poseidon2/kernels.cu.h
index 5f34be92..409098fc 100644
--- a/icicle/include/poseidon2/kernels.cu.h
+++ b/icicle/include/poseidon2/kernels.cu.h
@@ -3,10 +3,10 @@
 #define POSEIDON2_KERNELS_H
 
 #include "utils/utils.h"
-#include "hash/hash.cuh"
-#include "matrix/matrix.cuh"
-#include "poseidon2/constants.cuh"
-#include "gpu-utils/modifiers.cuh"
+#include "hash/hash.cu.h"
+#include "matrix/matrix.cu.h"
+#include "poseidon2/constants.cu.h"
+#include "gpu-utils/modifiers.cu.h"
 
 using matrix::Matrix;
 
diff --git a/icicle/include/poseidon2/poseidon2.cu.h b/icicle/include/poseidon2/poseidon2.cu.h
index 016740e7..c85e7d7c 100644
--- a/icicle/include/poseidon2/poseidon2.cu.h
+++ b/icicle/include/poseidon2/poseidon2.cu.h
@@ -4,15 +4,15 @@
 
 #include <cstdint>
 #include <stdexcept>
-#include "gpu-utils/device_context.cuh"
-#include "gpu-utils/error_handler.cuh"
+#include "gpu-utils/device_context.cu.h"
+#include "gpu-utils/error_handler.cu.h"
 #include "utils/utils.h"
 
-#include "hash/hash.cuh"
-#include "matrix/matrix.cuh"
+#include "hash/hash.cu.h"
+#include "matrix/matrix.cu.h"
 
-#include "poseidon2/constants.cuh"
-#include "poseidon2/kernels.cuh"
+#include "poseidon2/constants.cu.h"
+#include "poseidon2/kernels.cu.h"
 
 using matrix::Matrix;
 
diff --git a/icicle/include/vec_ops/vec_ops.cu.h b/icicle/include/vec_ops/vec_ops.cu.h
index ee624a07..00c4aac9 100644
--- a/icicle/include/vec_ops/vec_ops.cu.h
+++ b/icicle/include/vec_ops/vec_ops.cu.h
@@ -2,7 +2,7 @@
 #ifndef LDE_H
 #define LDE_H
 
-#include "gpu-utils/device_context.cuh"
+#include "gpu-utils/device_context.cu.h"
 
 /**
  * @namespace vec_ops
diff --git a/icicle/src/curves/extern.cu.cc b/icicle/src/curves/extern.cu.cc
index 8ea5bce2..d7384fa2 100644
--- a/icicle/src/curves/extern.cu.cc
+++ b/icicle/src/curves/extern.cu.cc
@@ -1,10 +1,10 @@
-#include "curves/curve_config.cuh"
+#include "curves/curve_config.cu.h"
 
 using namespace curve_config;
 
-#include "gpu-utils/device_context.cuh"
+#include "gpu-utils/device_context.cu.h"
 #include "utils/utils.h"
-#include "utils/mont.cuh"
+#include "utils/mont.cu.h"
 
 extern "C" bool CONCAT_EXPAND(CURVE, eq)(projective_t* point1, projective_t* point2)
 {
diff --git a/icicle/src/curves/extern_g2.cu.cc b/icicle/src/curves/extern_g2.cu.cc
index e8daa5e1..76f8a476 100644
--- a/icicle/src/curves/extern_g2.cu.cc
+++ b/icicle/src/curves/extern_g2.cu.cc
@@ -1,10 +1,10 @@
-#include "curves/curve_config.cuh"
+#include "curves/curve_config.cu.h"
 
 using namespace curve_config;
 
-#include "gpu-utils/device_context.cuh"
+#include "gpu-utils/device_context.cu.h"
 #include "utils/utils.h"
-#include "utils/mont.cuh"
+#include "utils/mont.cu.h"
 
 extern "C" bool CONCAT_EXPAND(CURVE, g2_eq)(g2_projective_t* point1, g2_projective_t* point2)
 {
diff --git a/icicle/src/fields/extern.cu.cc b/icicle/src/fields/extern.cu.cc
index 26c98ac3..7f65c228 100644
--- a/icicle/src/fields/extern.cu.cc
+++ b/icicle/src/fields/extern.cu.cc
@@ -1,10 +1,10 @@
-#include "fields/field_config.cuh"
+#include "fields/field_config.cu.h"
 
 using namespace field_config;
 
-#include "utils/mont.cuh"
+#include "utils/mont.cu.h"
 #include "utils/utils.h"
-#include "gpu-utils/device_context.cuh"
+#include "gpu-utils/device_context.cu.h"
 
 extern "C" void CONCAT_EXPAND(FIELD, generate_scalars)(scalar_t* scalars, int size)
 {
diff --git a/icicle/src/fields/extern_extension.cu.cc b/icicle/src/fields/extern_extension.cu.cc
index 8bc7bc5e..f8a1c822 100644
--- a/icicle/src/fields/extern_extension.cu.cc
+++ b/icicle/src/fields/extern_extension.cu.cc
@@ -1,10 +1,10 @@
-#include "fields/field_config.cuh"
+#include "fields/field_config.cu.h"
 
 using namespace field_config;
 
-#include "utils/mont.cuh"
+#include "utils/mont.cu.h"
 #include "utils/utils.h"
-#include "gpu-utils/device_context.cuh"
+#include "gpu-utils/device_context.cu.h"
 
 extern "C" void CONCAT_EXPAND(FIELD, extension_generate_scalars)(extension_t* scalars, int size)
 {
diff --git a/icicle/src/hash/keccak/extern.cu.cc b/icicle/src/hash/keccak/extern.cu.cc
index 009b8891..404ecac5 100644
--- a/icicle/src/hash/keccak/extern.cu.cc
+++ b/icicle/src/hash/keccak/extern.cu.cc
@@ -1,11 +1,11 @@
 #include "utils/utils.h"
-#include "gpu-utils/error_handler.cuh"
+#include "gpu-utils/error_handler.cu.h"
 
-#include "hash/hash.cuh"
-#include "hash/keccak/keccak.cuh"
+#include "hash/hash.cu.h"
+#include "hash/keccak/keccak.cu.h"
 #include "keccak.cu.cc"
 #include "../../merkle-tree/merkle.cu.cc"
-#include "merkle-tree/merkle.cuh"
+#include "merkle-tree/merkle.cu.h"
 
 namespace keccak {
   extern "C" cudaError_t
diff --git a/icicle/src/hash/keccak/keccak.cu.cc b/icicle/src/hash/keccak/keccak.cu.cc
index e805bcf6..ca3e52ba 100644
--- a/icicle/src/hash/keccak/keccak.cu.cc
+++ b/icicle/src/hash/keccak/keccak.cu.cc
@@ -1,10 +1,10 @@
 #include <cstdint>
-#include "gpu-utils/device_context.cuh"
-#include "gpu-utils/error_handler.cuh"
-#include "gpu-utils/modifiers.cuh"
+#include "gpu-utils/device_context.cu.h"
+#include "gpu-utils/error_handler.cu.h"
+#include "gpu-utils/modifiers.cu.h"
 
-#include "hash/hash.cuh"
-#include "hash/keccak/keccak.cuh"
+#include "hash/hash.cu.h"
+#include "hash/keccak/keccak.cu.h"
 
 using namespace hash;
 
diff --git a/icicle/src/hash/keccak/test.cu.cc b/icicle/src/hash/keccak/test.cu.cc
index fda294e4..2aac59a4 100644
--- a/icicle/src/hash/keccak/test.cu.cc
+++ b/icicle/src/hash/keccak/test.cu.cc
@@ -1,4 +1,4 @@
-#include "gpu-utils/device_context.cuh"
+#include "gpu-utils/device_context.cu.h"
 #include "extern.cu.cc"
 
 // #define DEBUG
diff --git a/icicle/src/hash/keccak/test_tree.cu.cc b/icicle/src/hash/keccak/test_tree.cu.cc
index 40b94af8..b4efe1c2 100644
--- a/icicle/src/hash/keccak/test_tree.cu.cc
+++ b/icicle/src/hash/keccak/test_tree.cu.cc
@@ -1,5 +1,5 @@
-#include "gpu-utils/device_context.cuh"
-#include "merkle-tree/merkle.cuh"
+#include "gpu-utils/device_context.cu.h"
+#include "merkle-tree/merkle.cu.h"
 #include "extern.cu.cc"
 
 #ifndef __CUDA_ARCH__
diff --git a/icicle/src/merkle-tree/extern.cu.cc b/icicle/src/merkle-tree/extern.cu.cc
index 6811473d..688a4366 100644
--- a/icicle/src/merkle-tree/extern.cu.cc
+++ b/icicle/src/merkle-tree/extern.cu.cc
@@ -1,12 +1,12 @@
 #include "utils/utils.h"
 
-#include "gpu-utils/error_handler.cuh"
-#include "merkle-tree/merkle.cuh"
+#include "gpu-utils/error_handler.cu.h"
+#include "merkle-tree/merkle.cu.h"
 #include "merkle.cu.cc"
 
-#include "hash/hash.cuh"
+#include "hash/hash.cu.h"
 
-#include "fields/field_config.cuh"
+#include "fields/field_config.cu.h"
 using namespace field_config;
 
 namespace merkle_tree {
diff --git a/icicle/src/merkle-tree/extern_mmcs.cu.cc b/icicle/src/merkle-tree/extern_mmcs.cu.cc
index 2d51dad1..3c2ba192 100644
--- a/icicle/src/merkle-tree/extern_mmcs.cu.cc
+++ b/icicle/src/merkle-tree/extern_mmcs.cu.cc
@@ -1,13 +1,13 @@
 #include "utils/utils.h"
 
-#include "gpu-utils/error_handler.cuh"
-#include "merkle-tree/merkle.cuh"
-#include "matrix/matrix.cuh"
+#include "gpu-utils/error_handler.cu.h"
+#include "merkle-tree/merkle.cu.h"
+#include "matrix/matrix.cu.h"
 #include "mmcs.cu.cc"
 
-#include "hash/hash.cuh"
+#include "hash/hash.cu.h"
 
-#include "fields/field_config.cuh"
+#include "fields/field_config.cu.h"
 using namespace field_config;
 
 using matrix::Matrix;
diff --git a/icicle/src/merkle-tree/merkle.cu.cc b/icicle/src/merkle-tree/merkle.cu.cc
index 53f3b8f6..1f783ab3 100644
--- a/icicle/src/merkle-tree/merkle.cu.cc
+++ b/icicle/src/merkle-tree/merkle.cu.cc
@@ -1,5 +1,5 @@
-#include "hash/hash.cuh"
-#include "merkle-tree/merkle.cuh"
+#include "hash/hash.cu.h"
+#include "merkle-tree/merkle.cu.h"
 
 namespace merkle_tree {
   /// Constructs merkle subtree without parallelization
diff --git a/icicle/src/merkle-tree/mmcs.cu.cc b/icicle/src/merkle-tree/mmcs.cu.cc
index f7acbb65..1d695483 100644
--- a/icicle/src/merkle-tree/mmcs.cu.cc
+++ b/icicle/src/merkle-tree/mmcs.cu.cc
@@ -1,7 +1,7 @@
-#include "hash/hash.cuh"
-#include "merkle-tree/merkle.cuh"
-#include "matrix/matrix.cuh"
-#include "vec_ops/vec_ops.cuh"
+#include "hash/hash.cu.h"
+#include "merkle-tree/merkle.cu.h"
+#include "matrix/matrix.cu.h"
+#include "vec_ops/vec_ops.cu.h"
 
 #include <algorithm>
 
diff --git a/icicle/src/merkle-tree/tests/merkle/test.cu.cc b/icicle/src/merkle-tree/tests/merkle/test.cu.cc
index 0fb34867..92401862 100644
--- a/icicle/src/merkle-tree/tests/merkle/test.cu.cc
+++ b/icicle/src/merkle-tree/tests/merkle/test.cu.cc
@@ -6,9 +6,9 @@
 #include <math.h>
 
 #define DEBUG
-#include "merkle-tree/merkle.cuh"
+#include "merkle-tree/merkle.cu.h"
 
-#include "poseidon/poseidon.cuh"
+#include "poseidon/poseidon.cu.h"
 
 #include "api/bls12_381.h"
 using namespace bls12_381;
diff --git a/icicle/src/merkle-tree/tests/merkle/test_poseidon2.cu.cc b/icicle/src/merkle-tree/tests/merkle/test_poseidon2.cu.cc
index 7bd35c92..c06a1bd2 100644
--- a/icicle/src/merkle-tree/tests/merkle/test_poseidon2.cu.cc
+++ b/icicle/src/merkle-tree/tests/merkle/test_poseidon2.cu.cc
@@ -5,9 +5,9 @@
 #include <iostream>
 #include <math.h>
 
-#include "merkle-tree/merkle.cuh"
+#include "merkle-tree/merkle.cu.h"
 
-#include "poseidon2/poseidon2.cuh"
+#include "poseidon2/poseidon2.cu.h"
 
 #include "api/babybear.h"
 using namespace babybear;
diff --git a/icicle/src/merkle-tree/tests/mmcs/test_poseidon2.cu.cc b/icicle/src/merkle-tree/tests/mmcs/test_poseidon2.cu.cc
index 9de240db..09434e63 100644
--- a/icicle/src/merkle-tree/tests/mmcs/test_poseidon2.cu.cc
+++ b/icicle/src/merkle-tree/tests/mmcs/test_poseidon2.cu.cc
@@ -5,9 +5,9 @@
 #include <iostream>
 #include <math.h>
 
-#include "merkle-tree/merkle.cuh"
+#include "merkle-tree/merkle.cu.h"
 
-#include "poseidon2/poseidon2.cuh"
+#include "poseidon2/poseidon2.cu.h"
 
 #include "api/babybear.h"
 using namespace babybear;
diff --git a/icicle/src/msm/extern.cu.cc b/icicle/src/msm/extern.cu.cc
index f4cb0ef3..42e05533 100644
--- a/icicle/src/msm/extern.cu.cc
+++ b/icicle/src/msm/extern.cu.cc
@@ -1,5 +1,5 @@
-#include "curves/curve_config.cuh"
-#include "fields/field_config.cuh"
+#include "curves/curve_config.cu.h"
+#include "fields/field_config.cu.h"
 
 using namespace curve_config;
 using namespace field_config;
diff --git a/icicle/src/msm/extern_g2.cu.cc b/icicle/src/msm/extern_g2.cu.cc
index 92b8f9fb..1a0d07b4 100644
--- a/icicle/src/msm/extern_g2.cu.cc
+++ b/icicle/src/msm/extern_g2.cu.cc
@@ -1,5 +1,5 @@
-#include "curves/curve_config.cuh"
-#include "fields/field_config.cuh"
+#include "curves/curve_config.cu.h"
+#include "fields/field_config.cu.h"
 
 using namespace curve_config;
 using namespace field_config;
diff --git a/icicle/src/msm/msm.cu.cc b/icicle/src/msm/msm.cu.cc
index 7b6f8851..6bb2139a 100644
--- a/icicle/src/msm/msm.cu.cc
+++ b/icicle/src/msm/msm.cu.cc
@@ -1,4 +1,4 @@
-#include "msm/msm.cuh"
+#include "msm/msm.cu.h"
 
 #include <cooperative_groups.h>
 #include <cub/device/device_radix_sort.cuh>
@@ -10,11 +10,11 @@
 #include <stdexcept>
 #include <vector>
 
-#include "curves/affine.cuh"
-#include "curves/projective.cuh"
-#include "fields/field.cuh"
-#include "gpu-utils/error_handler.cuh"
-#include "utils/mont.cuh"
+#include "curves/affine.cu.h"
+#include "curves/projective.cu.h"
+#include "fields/field.cu.h"
+#include "gpu-utils/error_handler.cu.h"
+#include "utils/mont.cu.h"
 
 namespace msm {
 
diff --git a/icicle/src/msm/tests/msm_test.cu.cc b/icicle/src/msm/tests/msm_test.cu.cc
index c3862fc9..81c82647 100644
--- a/icicle/src/msm/tests/msm_test.cu.cc
+++ b/icicle/src/msm/tests/msm_test.cu.cc
@@ -1,8 +1,8 @@
 #include "fields/id.h"
 // #define FIELD_ID 2
 #define CURVE_ID 3
-#include "curves/curve_config.cuh"
-// #include "fields/field_config.cuh"
+#include "curves/curve_config.cu.h"
+// #include "fields/field_config.cu.h"
 
 #include "msm.cu.cc"
 
@@ -10,10 +10,10 @@
 #include <iostream>
 #include <vector>
 
-#include "curves/params/bn254.cuh"
-#include "fields/field.cuh"
-#include "curves/projective.cuh"
-#include "gpu-utils/device_context.cuh"
+#include "curves/params/bn254.cu.h"
+#include "fields/field.cu.h"
+#include "curves/projective.cu.h"
+#include "gpu-utils/device_context.cu.h"
 
 // using namespace bn254;
 
diff --git a/icicle/src/ntt/extern.cu.cc b/icicle/src/ntt/extern.cu.cc
index 07291f1e..50a788ca 100644
--- a/icicle/src/ntt/extern.cu.cc
+++ b/icicle/src/ntt/extern.cu.cc
@@ -1,10 +1,10 @@
-#include "fields/field_config.cuh"
+#include "fields/field_config.cu.h"
 
 using namespace field_config;
 
 #include "ntt.cu.cc"
 
-#include "gpu-utils/device_context.cuh"
+#include "gpu-utils/device_context.cu.h"
 #include "utils/utils.h"
 
 namespace ntt {
diff --git a/icicle/src/ntt/extern_ecntt.cu.cc b/icicle/src/ntt/extern_ecntt.cu.cc
index 6469cf58..5cc0a64f 100644
--- a/icicle/src/ntt/extern_ecntt.cu.cc
+++ b/icicle/src/ntt/extern_ecntt.cu.cc
@@ -1,12 +1,12 @@
-#include "curves/curve_config.cuh"
-#include "fields/field_config.cuh"
+#include "curves/curve_config.cu.h"
+#include "fields/field_config.cu.h"
 
 using namespace curve_config;
 using namespace field_config;
 
 #include "ntt.cu.cc"
 
-#include "gpu-utils/device_context.cuh"
+#include "gpu-utils/device_context.cu.h"
 #include "utils/utils.h"
 
 namespace ntt {
diff --git a/icicle/src/ntt/extern_extension.cu.cc b/icicle/src/ntt/extern_extension.cu.cc
index d4431209..0c48adcc 100644
--- a/icicle/src/ntt/extern_extension.cu.cc
+++ b/icicle/src/ntt/extern_extension.cu.cc
@@ -1,10 +1,10 @@
-#include "fields/field_config.cuh"
+#include "fields/field_config.cu.h"
 
 using namespace field_config;
 
 #include "ntt.cu.cc"
 
-#include "gpu-utils/device_context.cuh"
+#include "gpu-utils/device_context.cu.h"
 #include "utils/utils.h"
 
 namespace ntt {
diff --git a/icicle/src/ntt/kernel_ntt.cu.cc b/icicle/src/ntt/kernel_ntt.cu.cc
index 93f073c4..1314a04f 100644
--- a/icicle/src/ntt/kernel_ntt.cu.cc
+++ b/icicle/src/ntt/kernel_ntt.cu.cc
@@ -1,10 +1,10 @@
-#include "fields/field_config.cuh"
+#include "fields/field_config.cu.h"
 
 using namespace field_config;
 
 #include "thread_ntt.cu.cc"
-#include "gpu-utils/sharedmem.cuh"
-#include "ntt/ntt.cuh" // for ntt::Ordering
+#include "gpu-utils/sharedmem.cu.h"
+#include "ntt/ntt.cu.h" // for ntt::Ordering
 
 namespace mxntt {
 
diff --git a/icicle/src/ntt/ntt.cu.cc b/icicle/src/ntt/ntt.cu.cc
index 0afe4a29..a82522a0 100644
--- a/icicle/src/ntt/ntt.cu.cc
+++ b/icicle/src/ntt/ntt.cu.cc
@@ -1,22 +1,22 @@
-#include "fields/field_config.cuh"
+#include "fields/field_config.cu.h"
 
 using namespace field_config;
 
-#include "ntt/ntt.cuh"
+#include "ntt/ntt.cu.h"
 
 #include <unordered_map>
 #include <vector>
 #include <type_traits>
 
-#include "gpu-utils/sharedmem.cuh"
-#include "utils/utils_kernels.cuh"
+#include "gpu-utils/sharedmem.cu.h"
+#include "utils/utils_kernels.cu.h"
 #include "utils/utils.h"
-#include "ntt/ntt_impl.cuh"
+#include "ntt/ntt_impl.cu.h"
 
 #include <mutex>
 
 #ifdef CURVE_ID
-#include "curves/curve_config.cuh"
+#include "curves/curve_config.cu.h"
 using namespace curve_config;
 #define IS_ECNTT std::is_same_v<E, projective_t>
 #else
diff --git a/icicle/src/ntt/tests/verification.cu.cc b/icicle/src/ntt/tests/verification.cu.cc
index edb769a8..ddf846b9 100644
--- a/icicle/src/ntt/tests/verification.cu.cc
+++ b/icicle/src/ntt/tests/verification.cu.cc
@@ -3,17 +3,17 @@
 
 #ifdef ECNTT
 #define CURVE_ID BN254
-#include "curves/curve_config.cuh"
+#include "curves/curve_config.cu.h"
 typedef field_config::scalar_t test_scalar;
 typedef curve_config::projective_t test_data;
 #else
-#include "fields/field_config.cuh"
+#include "fields/field_config.cu.h"
 typedef field_config::scalar_t test_scalar;
 typedef field_config::scalar_t test_data;
 #endif
 
-#include "fields/field.cuh"
-#include "curves/projective.cuh"
+#include "fields/field.cu.h"
+#include "curves/projective.cu.h"
 #include <chrono>
 #include <iostream>
 #include <vector>
diff --git a/icicle/src/ntt/thread_ntt.cu.cc b/icicle/src/ntt/thread_ntt.cu.cc
index 8fd2764d..dd7fd7b6 100644
--- a/icicle/src/ntt/thread_ntt.cu.cc
+++ b/icicle/src/ntt/thread_ntt.cu.cc
@@ -4,7 +4,7 @@
 
 #include <stdio.h>
 #include <stdint.h>
-#include "gpu-utils/modifiers.cuh"
+#include "gpu-utils/modifiers.cu.h"
 
 struct stage_metadata {
   uint32_t th_stride;
diff --git a/icicle/src/polynomials/cuda_backend/polynomial_cuda_backend.cu.cc b/icicle/src/polynomials/cuda_backend/polynomial_cuda_backend.cu.cc
index 1b9d5d14..d3fef443 100644
--- a/icicle/src/polynomials/cuda_backend/polynomial_cuda_backend.cu.cc
+++ b/icicle/src/polynomials/cuda_backend/polynomial_cuda_backend.cu.cc
@@ -1,11 +1,11 @@
 
 #include "polynomials/polynomials.h"
-#include "polynomials/cuda_backend/polynomial_cuda_backend.cuh"
+#include "polynomials/cuda_backend/polynomial_cuda_backend.cu.h"
 
-#include "gpu-utils/device_context.cuh"
+#include "gpu-utils/device_context.cu.h"
 #include "cuda_runtime.h"
-#include "ntt/ntt.cuh"
-#include "kernels.cuh"
+#include "ntt/ntt.cu.h"
+#include "kernels.cu.h"
 
 using device_context::DeviceContext;
 
diff --git a/icicle/src/polynomials/polynomials_c_api.cu.cc b/icicle/src/polynomials/polynomials_c_api.cu.cc
index 8672fbd9..b9066c4f 100644
--- a/icicle/src/polynomials/polynomials_c_api.cu.cc
+++ b/icicle/src/polynomials/polynomials_c_api.cu.cc
@@ -1,8 +1,8 @@
 #include "polynomials/polynomials.h"
-#include "fields/field_config.cuh"
+#include "fields/field_config.cu.h"
 #include "utils/utils.h"
 #include "utils/integrity_pointer.h"
-#include "polynomials/cuda_backend/polynomial_cuda_backend.cuh"
+#include "polynomials/cuda_backend/polynomial_cuda_backend.cu.h"
 
 namespace polynomials {
   extern "C" {
diff --git a/icicle/src/poseidon/constants.cu.cc b/icicle/src/poseidon/constants.cu.cc
index bed5a9be..95a51673 100644
--- a/icicle/src/poseidon/constants.cu.cc
+++ b/icicle/src/poseidon/constants.cu.cc
@@ -1,5 +1,5 @@
-#include "poseidon/constants.cuh"
-#include "gpu-utils/device_context.cuh"
+#include "poseidon/constants.cu.h"
+#include "gpu-utils/device_context.cu.h"
 
 /// These are pre-calculated constants for different curves
 #include "fields/id.h"
diff --git a/icicle/src/poseidon/extern.cu.cc b/icicle/src/poseidon/extern.cu.cc
index 2d7ee1e5..b10db1d8 100644
--- a/icicle/src/poseidon/extern.cu.cc
+++ b/icicle/src/poseidon/extern.cu.cc
@@ -1,11 +1,11 @@
-#include "fields/field_config.cuh"
+#include "fields/field_config.cu.h"
 
 using namespace field_config;
 
-#include "poseidon/poseidon.cuh"
+#include "poseidon/poseidon.cu.h"
 #include "constants.cu.cc"
 
-#include "gpu-utils/device_context.cuh"
+#include "gpu-utils/device_context.cu.h"
 #include "utils/utils.h"
 
 namespace poseidon {
diff --git a/icicle/src/poseidon/test.cu.cc b/icicle/src/poseidon/test.cu.cc
index 461b57ae..4384da3d 100644
--- a/icicle/src/poseidon/test.cu.cc
+++ b/icicle/src/poseidon/test.cu.cc
@@ -1,9 +1,9 @@
 // #define DEBUG
 
-#include "curves/curve_config.cuh"
+#include "curves/curve_config.cu.h"
 using namespace curve_config;
 
-#include "gpu-utils/device_context.cuh"
+#include "gpu-utils/device_context.cu.h"
 
 #ifndef __CUDA_ARCH__
 #include <cassert>
@@ -14,7 +14,7 @@ using namespace curve_config;
 #include "api/bls12_381.h"
 using namespace bls12_381;
 
-#include "poseidon/poseidon.cuh"
+#include "poseidon/poseidon.cu.h"
 using namespace poseidon;
 
 #define A 2
diff --git a/icicle/src/poseidon/test_m31.cu.cc b/icicle/src/poseidon/test_m31.cu.cc
index 7f6bd7ce..cbf22de7 100644
--- a/icicle/src/poseidon/test_m31.cu.cc
+++ b/icicle/src/poseidon/test_m31.cu.cc
@@ -1,10 +1,10 @@
 // #define DEBUG
 
-#include "fields/field_config.cuh"
+#include "fields/field_config.cu.h"
 using namespace field_config;
 
-#include "gpu-utils/device_context.cuh"
-#include "poseidon/poseidon.cuh"
+#include "gpu-utils/device_context.cu.h"
+#include "poseidon/poseidon.cu.h"
 
 #ifndef __CUDA_ARCH__
 #include <cassert>
diff --git a/icicle/src/poseidon2/constants.cu.cc b/icicle/src/poseidon2/constants.cu.cc
index 3fa157d2..c30fab04 100644
--- a/icicle/src/poseidon2/constants.cu.cc
+++ b/icicle/src/poseidon2/constants.cu.cc
@@ -1,5 +1,5 @@
-#include "poseidon2/constants.cuh"
-#include "gpu-utils/device_context.cuh"
+#include "poseidon2/constants.cu.h"
+#include "gpu-utils/device_context.cu.h"
 
 /// These are pre-calculated constants for different curves
 #include "fields/id.h"
diff --git a/icicle/src/poseidon2/extern.cu.cc b/icicle/src/poseidon2/extern.cu.cc
index 50b9bdcb..d2397255 100644
--- a/icicle/src/poseidon2/extern.cu.cc
+++ b/icicle/src/poseidon2/extern.cu.cc
@@ -1,10 +1,10 @@
 #include "utils/utils.h"
 
-#include "fields/field_config.cuh"
+#include "fields/field_config.cu.h"
 using namespace field_config;
 
-#include "gpu-utils/error_handler.cuh"
-#include "poseidon2/poseidon2.cuh"
+#include "gpu-utils/error_handler.cu.h"
+#include "poseidon2/poseidon2.cu.h"
 #include "./constants.cu.cc"
 
 namespace poseidon2 {
diff --git a/icicle/src/poseidon2/test.cu.cc b/icicle/src/poseidon2/test.cu.cc
index c98bfd63..91e8da8c 100644
--- a/icicle/src/poseidon2/test.cu.cc
+++ b/icicle/src/poseidon2/test.cu.cc
@@ -1,4 +1,4 @@
-#include "gpu-utils/device_context.cuh"
+#include "gpu-utils/device_context.cu.h"
 
 #ifndef __CUDA_ARCH__
 #include <cassert>
@@ -6,7 +6,7 @@
 #include <fstream>
 #include <iostream>
 
-#include "poseidon2/poseidon2.cuh"
+#include "poseidon2/poseidon2.cu.h"
 using namespace poseidon2;
 
 #include "api/bn254.h"
diff --git a/icicle/src/poseidon2/test_m31.cu.cc b/icicle/src/poseidon2/test_m31.cu.cc
index 378d281f..ed77de1f 100644
--- a/icicle/src/poseidon2/test_m31.cu.cc
+++ b/icicle/src/poseidon2/test_m31.cu.cc
@@ -1,4 +1,4 @@
-#include "gpu-utils/device_context.cuh"
+#include "gpu-utils/device_context.cu.h"
 
 #ifndef __CUDA_ARCH__
 #include <cassert>
@@ -6,13 +6,13 @@
 #include <fstream>
 #include <iostream>
 
-#include "poseidon2/poseidon2.cuh"
+#include "poseidon2/poseidon2.cu.h"
 using namespace poseidon2;
 
-#include "fields/field_config.cuh"
+#include "fields/field_config.cu.h"
 using namespace field_config;
 
-#include "hash/hash.cuh"
+#include "hash/hash.cu.h"
 
 #define T 16
 
diff --git a/icicle/src/vec_ops/extern.cu.cc b/icicle/src/vec_ops/extern.cu.cc
index 21b2c071..f715fd83 100644
--- a/icicle/src/vec_ops/extern.cu.cc
+++ b/icicle/src/vec_ops/extern.cu.cc
@@ -1,4 +1,4 @@
-#include "fields/field_config.cuh"
+#include "fields/field_config.cu.h"
 
 using namespace field_config;
 
diff --git a/icicle/src/vec_ops/extern_extension.cu.cc b/icicle/src/vec_ops/extern_extension.cu.cc
index 5c0fce48..30d12147 100644
--- a/icicle/src/vec_ops/extern_extension.cu.cc
+++ b/icicle/src/vec_ops/extern_extension.cu.cc
@@ -1,4 +1,4 @@
-#include "fields/field_config.cuh"
+#include "fields/field_config.cu.h"
 
 using namespace field_config;
 
diff --git a/icicle/src/vec_ops/vec_ops.cu.cc b/icicle/src/vec_ops/vec_ops.cu.cc
index 9cfa85e6..c813dafa 100644
--- a/icicle/src/vec_ops/vec_ops.cu.cc
+++ b/icicle/src/vec_ops/vec_ops.cu.cc
@@ -1,9 +1,9 @@
 #include <cuda.h>
 #include <stdexcept>
 
-#include "vec_ops/vec_ops.cuh"
-#include "gpu-utils/device_context.cuh"
-#include "utils/mont.cuh"
+#include "vec_ops/vec_ops.cu.h"
+#include "gpu-utils/device_context.cu.h"
+#include "utils/mont.cu.h"
 
 namespace vec_ops {
 
diff --git a/icicle/tests/curve_test.cu.cc b/icicle/tests/curve_test.cu.cc
index 087b3e4b..6b122722 100644
--- a/icicle/tests/curve_test.cu.cc
+++ b/icicle/tests/curve_test.cu.cc
@@ -1,5 +1,5 @@
-#include "utils/test_functions.cuh"
-#include "curves/curve_config.cuh"
+#include "utils/test_functions.cu.h"
+#include "curves/curve_config.cu.h"
 #include <cuda_runtime.h>
 #include <gtest/gtest.h>
 #include <iostream>
diff --git a/icicle/tests/device_error_test.cu.cc b/icicle/tests/device_error_test.cu.cc
index 9071f883..44b3a0bf 100644
--- a/icicle/tests/device_error_test.cu.cc
+++ b/icicle/tests/device_error_test.cu.cc
@@ -1,4 +1,4 @@
-#include "gpu-utils/error_handler.cuh" // Include your error handling header file
+#include "gpu-utils/error_handler.cu.h" // Include your error handling header file
 #include <gtest/gtest.h>
 
 __global__ void a_kernel_with_conditional_sticky_error(bool is_failing)
diff --git a/icicle/tests/error_handler_test.cu.cc b/icicle/tests/error_handler_test.cu.cc
index 64023694..5a88865d 100644
--- a/icicle/tests/error_handler_test.cu.cc
+++ b/icicle/tests/error_handler_test.cu.cc
@@ -1,4 +1,4 @@
-#include "gpu-utils/error_handler.cuh" // Include your error handling header file
+#include "gpu-utils/error_handler.cu.h" // Include your error handling header file
 #include <gtest/gtest.h>
 
 class IcicleErrorTest : public ::testing::Test
diff --git a/icicle/tests/field_test.cu.cc b/icicle/tests/field_test.cu.cc
index 704c3986..ed30c3e5 100644
--- a/icicle/tests/field_test.cu.cc
+++ b/icicle/tests/field_test.cu.cc
@@ -1,5 +1,5 @@
-#include "utils/test_functions.cuh"
-#include "fields/field_config.cuh"
+#include "utils/test_functions.cu.h"
+#include "fields/field_config.cu.h"
 #include <cuda_runtime.h>
 #include <gtest/gtest.h>
 #include <iostream>
diff --git a/icicle/tests/polynomial_test.cu.cc b/icicle/tests/polynomial_test.cu.cc
index 9b433fa1..55c9942e 100644
--- a/icicle/tests/polynomial_test.cu.cc
+++ b/icicle/tests/polynomial_test.cu.cc
@@ -5,10 +5,10 @@
 #include <list>
 
 #include "polynomials/polynomials.h"
-#include "polynomials/cuda_backend/polynomial_cuda_backend.cuh"
+#include "polynomials/cuda_backend/polynomial_cuda_backend.cu.h"
 
-#include "ntt/ntt.cuh"
-#include "gpu-utils/device_context.cuh"
+#include "ntt/ntt.cu.h"
+#include "gpu-utils/device_context.cu.h"
 
 /*******************************************/
 
@@ -516,8 +516,8 @@ TEST_F(PolynomialTest, slicing)
 }
 
 #ifdef CURVE
-#include "msm/msm.cuh"
-#include "curves/curve_config.cuh"
+#include "msm/msm.cu.h"
+#include "curves/curve_config.cu.h"
 using curve_config::affine_t;
 using curve_config::projective_t;
 
